besskge/pipeline.py

```diff
@@ -0,0 +1,298 @@
+        :param batch_sampler:
+            Batch sampler, based on a
+            "h_shard"/"t_shard"-partitioned triple set.
+        :param corruption_scheme:
+            Set to "t" to score (h, r, ?) completions, or to
+            "h" to score (?, r, t) completions.
+        :param score_fn:
+            The trained scoring function.
+        :param evaluation:
+            Evaluation module, for computing metrics.
+            Default: None.
+        :param filter_triples:
+            The set of all triples whose scores need to be filtered.
+            The triples passed here must have GLOBAL IDs for head/tail
+            entities. Default: None.
+        :param return_scores:
+            If True, store and return scores of all queries' completions
+            (with filters applied, if specified).
+            For large number of queries/entities, this can cause the host
+            to go OOM.
+            Default: False.
+        :param return_topk:
+            If True, return for each query the global IDs of the most likely
+            completions, after filtering out the scores of `filter_triples`.
+            Default: False.
+        :param k:
+            If `return_topk` is set to True, for each query return the
+            top-k most likely predictions (after filtering). Default: 10.
+        :param windows_size:
+            Size of the sliding window, namely the number of negative entities
+            scored against each query at each step on IPU and returned to host.
+            Should be decreased with large batch sizes, to avoid an OOM error.
+            Default: 1000.
+        :param use_ipu_model:
+            Run pipeline on IPU Model instead of actual hardware. Default: False.
+        """
+        super().__init__()
+        self.batch_sampler = batch_sampler
+        if not (evaluation or return_scores):
+            raise ValueError(
+                "Nothing to return. Provide `evaluation` or set `return_scores=True`"
+            )
+        if corruption_scheme not in ["h", "t"]:
+            raise ValueError("corruption_scheme needs to be either 'h' or 't'")
+        if (
+            corruption_scheme == "h"
+            and self.batch_sampler.triple_partition_mode != "t_shard"
+        ):
+            raise ValueError(
+                "Corruption scheme 'h' requires 't-shard'-partitioned triples"
+            )
+        elif (
+            corruption_scheme == "t"
+            and self.batch_sampler.triple_partition_mode != "h_shard"
+        ):
+            raise ValueError(
+                "Corruption scheme 't' requires 'h-shard'-partitioned triples"
+            )
+        self.candidate_sampler = PlaceholderNegativeSampler(
+            corruption_scheme=corruption_scheme
+        )
+        self.score_fn = score_fn
+        self.evaluation = evaluation
+        self.return_scores = return_scores
+        self.return_topk = return_topk
+        self.k = k
+        self.window_size = windows_size
```

Comment:
One more random (and very minor) question: Any particular reason why the one side is `window_size` and the other is `windows_size` (The reason I'm asking is that I just ran some experiments and expected to pass `window_size` as argument)

Reply:
My bad, that's a typo. Thanks for catching this, fixed in [d30fff8](https://github.com/graphcore-research/bess-kge/pull/33/commits/d30fff80be7b2188b676db4f470637b0d5fadd43)